src/eos-updater-util.c

```diff
@@ -313,73 +323,180 @@ eos_updater_resolve_upgrade (EosUpdater  *updater,
   bkf = g_key_file_new ();
   if (!g_key_file_load_from_data (bkf, msg->response_body->data, -1,
                                   G_KEY_FILE_NONE, error))
-    goto out;
+    return NULL;
+  return g_steal_pointer (&bkf);
+}
+
+static gboolean
+process_single_group (GKeyFile *bkf,
+                      const gchar *group_name,
+                      gboolean *on_hold,
+                      gchar **p_ref,
+                      GError **error)
+{
+  if (g_key_file_get_boolean (bkf, group_name, ON_HOLD_KEY, NULL))
+    {
+      *on_hold = TRUE;
+      return TRUE;
+    }
+  *p_ref = g_key_file_get_string (bkf, group_name, OSTREE_REF_KEY, error);
+  if (*p_ref == NULL)
+    return FALSE;
+  return TRUE;
+}
 
+static gboolean
+process_branch_file (GKeyFile *bkf,
+                     const gchar *group_name,
+                     gboolean *on_hold,
+                     gchar **p_ref,
+                     GError **error)
+{
   /* Check for product-specific entry */
-  if (g_key_file_has_group (bkf, p_id))
+  if (g_key_file_has_group (bkf, group_name))
     {
       message ("Product-specific branch configuration found");
-      if (g_key_file_get_boolean (bkf, p_id, ON_HOLD_KEY, NULL))
-        {
-          message ("Product is on hold, nothing to upgrade here");
-          on_hold = TRUE;
-          ret = TRUE;
-          goto out;
-        }
-      p_ref = g_key_file_get_string (bkf, p_id, OSTREE_REF_KEY, error);
-      if (!p_ref)
-        goto out;
+      if (!process_single_group (bkf, group_name, on_hold, p_ref, error))
+        return FALSE;
+      if (*on_hold)
+        message ("Product is on hold, nothing to upgrade here");
+      return TRUE;
     }
   /* Check for a DEFAULT_GROUP entry */
-  else if (g_key_file_has_group (bkf, DEFAULT_GROUP))
+  if (g_key_file_has_group (bkf, DEFAULT_GROUP))
     {
       message ("No product-specific branch configuration found, following %s",
                DEFAULT_GROUP);
-      if (g_key_file_get_boolean (bkf, DEFAULT_GROUP, ON_HOLD_KEY, NULL))
-        {
-          message ("No product-specific configuration and %s is on hold, "
-                   "nothing to upgrade here", DEFAULT_GROUP);
-          on_hold = TRUE;
-          ret = TRUE;
-          goto out;
-        }
-      p_ref = g_key_file_get_string (bkf, DEFAULT_GROUP, OSTREE_REF_KEY, error);
-      if (!p_ref)
-        goto out;
+      if (!process_single_group (bkf, group_name, on_hold, p_ref, error))
+        return FALSE;
+      if (*on_hold)
+        message ("No product-specific configuration and %s is on hold, "
+                 "nothing to upgrade here", DEFAULT_GROUP);
+      return TRUE;
     }
-  else /* fallback to the the origin file ref */
+  *p_ref = NULL;
+  return TRUE;
+}
+
+static void
+maybe_send_metric (const gchar *vendor,
+                   const gchar *product,
+                   const gchar *ref,
+                   gboolean on_hold)
+{
+  static gboolean metric_sent = FALSE;
+
+  if (G_LIKELY (metric_sent))
```

Comment:
We usually don't use `G_LIKELY` and `G_UNLIKELY` because they make the code more complicated and their usefulness is questionable... Please remove